sql/expression/function/aggregation/sum.go

```diff
@@ -0,0 +1,144 @@
+package aggregation
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/davecgh/go-spew/spew"
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+	"gopkg.in/src-d/go-mysql-server.v0/sql/expression"
+)
+
+// Sum agregation returns the sum of all values in the selected column.
+// It implements the Aggregation interface.
+type Sum struct {
+	expression.UnaryExpression
+}
+
+// NewSum returns a new Sum node.
+func NewSum(e sql.Expression) *Sum {
+	return &Sum{expression.UnaryExpression{Child: e}}
+}
+
+// Resolved implements the Resolvable interface.
+func (m *Sum) Resolved() bool {
+	return m.Child.Resolved()
+}
+
+// Type returns the resultant type of the aggregation.
+func (m *Sum) Type() sql.Type {
+	return sql.Int64
+}
+
+func (m *Sum) String() string {
+	return fmt.Sprintf("SUM(%s)", m.Child)
+}
+
+// IsNullable returns whether the return value can be null.
+func (m *Sum) IsNullable() bool {
+	return false
+}
+
+// TransformUp implements the Transformable interface.
+func (m *Sum) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	child, err := m.Child.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+	return f(NewSum(child))
+}
+
+// NewBuffer creates a new buffer to compute the result.
+func (m *Sum) NewBuffer() sql.Row {
+	return sql.NewRow(nil)
+}
+
+func isFloatSum(buffer sql.Row, value interface{}) bool {
+	if _, ok := buffer[0].(float64); ok {
+		return true
+	}
+
+	fv, err := sql.Float64.Convert(value)
+	if err != nil {
+		return false
+	}
+
+	// if can be converted from float64 but not int64 it's a float sum
+	iv, err := sql.Int64.Convert(value)
+	if err != nil {
+		return true
+	}
+
+	// they're different, so that means float got casted to
+	// int and lost the fraction part
+	if float64(iv.(int64)) != fv.(float64) {
+		return true
+	}
+
+	return false
+}
+
+// Update implements the Aggregation interface.
+func (m *Sum) Update(ctx *sql.Context, buffer, row sql.Row) error {
+	v, err := m.Child.Eval(ctx, row)
+	if err != nil {
+		return err
+	}
+
+	if v == nil {
+		return nil
+	}
+
+	if isFloatSum(buffer, v) {
+		fv, err := sql.Float64.Convert(v)
+		if err != nil {
+			fv = float64(0)
+		}
+
+		switch b := buffer[0].(type) {
+		case nil:
+			buffer[0] = fv.(float64)
+		case int64:
+			buffer[0] = float64(b) + fv.(float64)
+		case float64:
+			buffer[0] = b + fv.(float64)
+		default:
+			return sql.ErrInvalidType.New(reflect.TypeOf(b))
```

Comment:
Sum shouldn't throw an error on unsupported types. `SELECT SUM('test')` returns 0

Reply:
When is this error thrown? I check that you have tests for strings too.

Reply:
This is for checking the type of the internal buffer. This error should never be returned, it would mean there's a bug in the code. Perhaps it would be better to change it to another error like `this is likely a bug, detected value of type foo in the aggregation buffer`?